src/Compiler/Microsoft.CodeAnalysis.Razor/src/RenderModeTagHelperDescriptorProvider.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using static Microsoft.AspNetCore.Razor.Language.CommonMetadata;
+
+namespace Microsoft.CodeAnalysis.Razor;
+
+internal class RenderModeTagHelperDescriptorProvider : ITagHelperDescriptorProvider
+{
+    private static TagHelperDescriptor? s_refTagHelper;
+
+    // Run after the component tag helper provider
+    public int Order { get; set; } = 1000;
+
+    public RazorEngine? Engine { get; set; }
+
+    public void Execute(TagHelperDescriptorProviderContext context)
+    {
+        if (context == null)
+        {
+            throw new ArgumentNullException(nameof(context));
+        }
+
+        var compilation = context.GetCompilation();
+        if (compilation == null)
+        {
+            return;
+        }
+
+        var iComponentRenderMode = compilation.GetTypeByMetadataName(ComponentsApi.IComponentRenderMode.FullTypeName);
+        if (iComponentRenderMode == null)
+        {
+            // If we can't find IComponentRenderMode, then just bail. We won't be able to compile the
+            // generated code anyway.
+            return;
+        }
+
+        var targetSymbol = context.Items.GetTargetSymbol();
+        if (targetSymbol is not null && !SymbolEqualityComparer.Default.Equals(targetSymbol, iComponentRenderMode.ContainingAssembly))
```

Comment:
What is this check doing?

Reply:
When `targetSymbol` is set, it means a specific assembly is being searched for tag helpers. If the assembly where `iComponentRenderMode` is defined is different from the assembly being searched, then don't add the tag helper (as we could end up with it multiple times).

Essentially, we only add it if we're searching the assembly where it would exist if it existed in source.